geomet/geopackage.py

```diff
@@ -0,0 +1,394 @@
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import struct as _struct
+
+from geomet.util import (
+    as_bin_str as _as_bin_str,
+    take as _take,
+    endian_token as _endian_token
+)
+from geomet import wkb as _wkb
+
+
+def dump(obj, dest_file, big_endian=True):
+    """
+    Dump GeoJSON-like `dict` to GeoPackage binary
+    and write it to the `dest_file`.
+
+    :param dict obj:
+        A GeoJSON-like dictionary. It must at least the keys 'type' and
+        'coordinates'.
+    :param dest_file:
+        Open and writable file-like object.
+    :param bool big_endian:
+        specify endianess of the dumped object.
+
+    :return:
+    """
+    dest_file.write(dumps(obj, big_endian))
+
+
+def load(source_file):
+    """
+    Load a GeoJSON `dict` object from a ``source_file`` containing
+    GeoPackage (as a byte string).
+
+    :param source_file:
+        Open and readable file-like object.
+
+    :return:
+        A GeoJSON `dict` representing the geometry read from the file.
+    """
+    return loads(source_file.read())
+
+
+def dumps(obj, big_endian=True):
+    """
+    Dump a GeoJSON-like dict to a GeoPackage bytestring.
+
+
+    If the dict contains a top-level 'meta' key like so:
+
+    ```
+    'meta': {'srid': 4326}
+    ```
+    then the srid will be added to the geopackage header, but *not*
+    to the WKB geometry header.
+
+
+    If the dict contains a top-level 'bbox' key like so:
+
+    ```
+    'bbox': [0, 0, 3, 3]
+    ```
+
+    Then an envelope will be added to the geopackage header
+    with this information.
+
+
+    If the geometry's coordinates are empty (an empty list)
+    then the geopackage header's "empty" flag will be set,
+    denoting that this geometry has no coordinates.
+
+    Please note that while this library can parse geopackages
+    with a mixed byte-order in the header, it will only produce
+    blobs with consistent byte order (albeit properly marked as such).
+    That means you cannot product a geopackage with e.g. little-endian
+    header and big-endian WKB geometry.
+
+    :param dict obj:
+        The geojson geometry to dump
+    :param bool big_endian:
+        if True, the geopackage binary will use big-endian
+        byte order, little-endian otherwise.
+
+    :return bytes:
+        bytestring representing the geometry in geopackage
+        format.
+    """
+    header = _build_geopackage_header(obj, not big_endian)
+    result = _wkb._dumps(obj, big_endian, include_meta=False)
+    return header + result
+
+
+def loads(string):
+    """
+    Construct a GeoJSON `dict` from geopackage (string).
+
+    This function strips the geopackage header from the
+    string and passes the WKB after it directly to the
+    `geomet.wkb.loads` function.
+
+    The envelope, if present, is added to the GeoJSON as
+    a key called 'bbox' as per the GeoJSON spec, [1].
```

Comment:
Hey @larsbutler, there is a little "gotcha" there, though. The problem is that `bbox` is a part of feature-level spec, not the geometry, so while we're adding an item which is part of the spec it's not in the coreect place, obviously because geomet doesn't deal in features. I posted a [comment in the issue](https://github.com/geomet/geomet/issues/42#issuecomment-599824385) for geopackage which discusses the options. While I went with a root-level key on the geometry, I'm open to suggestions and changing it would be minor.

Reply:
That's fine; let's go with this for now and if there's a compelling reason to change it later, we can do that.